### ISSUE TITLE
Does draft distillation actually raise acceptance? (7B+ / genuinely diverged fine-tune)

### ISSUE THREAD

**Issue by MakazhanAlpamys** [MakazhanAlpamys/Soup#303]
## The honest state

v0.71.33 shipped `soup draft` (distil + **measure** a speculative-decoding draft). The measurement half is solid. The **distillation half's central premise is unproven** — and on the only pair we could validate on a 4 GB box, it did **not** pay off.

Measured live on an RTX 3050, `HuggingFaceTB/SmolLM2-360M-Instruct` (target) ← `HuggingFaceTB/SmolLM2-135M-Instruct` (draft):

| draft | acceptance | assisted speedup |
|---|---|---|
| stock, no distillation | **69.3%** | 0.58× |
| distilled, 2 epochs (264 steps) | 69.7% | 0.55× |
| distilled, **10 epochs** (1320 steps) | **69.3%** | 0.62× |

5× more training moved acceptance by **0.0 points**. Assisted decoding was a **net slowdown** throughout.

Our reading: a small **same-family** draft is already near its capacity ceiling for agreeing with the target. The tokens it gets wrong are the ones it lacks the capacity/knowledge to get right, and logit KD can't buy capacity that isn't there. But this is a hypothesis derived from one small pair — we can't distinguish it from "our distill recipe is too weak" or "the regime is wrong" on the hardware we have.

## What we're asking

If you have a bigger GPU, please check whether draft distillation raises acceptance where we predict it *should*:

1. **A genuinely diverged target.** A model fine-tuned hard on a domain/style/structured-output format, so that off-the-shelf drafts no longer predict it. (We tried to construct one on 4 GB; our small target either overfit into degenerate repetition — trivially predictable, α went *up* to 70.3% — or didn't learn the style at all. Both are artifacts of the size, not of the idea.)
2. **A real-scale pair.** e.g. Llama-3.1-70B ← Llama-3.2-1B, or Qwen2.5-32B ← Qwen2.5-0.5B, where the draft is ~1-2% of the target and assisted decoding actually has room to win.

## Reproduce

```bash
# 1. baseline: how good is the stock draft already?
soup draft measure --target <your-tuned-model> --draft <tiny-same-tokenizer-model> \
  --prompts prod-prompts.jsonl -o before.json

# 2. distil a draft from your target's own distribution
soup draft distill --target <your-tuned-model> --draft-base <tiny-same-tokenizer-model> \
  --data traffic.jsonl --steps 2000 -o draft/

# 3. did it move?
soup draft measure --target <your-tuned-model> --draft draft/ \
  --prompts prod-prompts.jsonl -o after.json
```

## Acceptance criteria

- [ ] A metric-logged before/after on a ≥7B target (or a genuinely diverged fine-tune) showing whether acceptance rises, with `before.json` / `after.json` attached.
- [ ] The corresponding **wall-clock** plain-vs-assisted tok/s, so we learn where speculative decoding actually starts winning.
- [ ] **Either** outcome is a valid, wanted answer. A documented failure ("distillation doesn't help at scale either, here's the data") is as useful as a success and we will ship it in the docs.

Related: #286 (the same scale ask for the reward-hacking controller). Both exist because the maintainer's box is a single RTX 3050 and we would rather say so than publish an unvalidated speedup number.

**Comment by MakazhanAlpamys** [comment]
Answered on an H100. **Distillation does not raise acceptance at real scale either — and for this pair acceptance was never the binding constraint.**

## The null result reproduces at 8B

Llama-3.1-8B-Instruct ← Llama-3.2-1B-Instruct (matching vocab, `same_tokenizer` passes), 48 fixed prompts, distillation corpus = 400 **disjoint** prompts answered by the target itself — the recipe most favourable to the feature.

| | acceptance | plain tok/s | assisted tok/s | speedup |
|---|---|---|---|---|
| before | **0.81293** | 39.28 | 18.90 | 0.481 |
| after | **0.81255** | 52.00 | 28.19 | 0.542 |

**Δ = −0.038 percentage points.** Paired instrument, both drafts scored against the *same* target continuations: Δ = −0.000378, 95% CI **[−0.0111, +0.0107]**. Token contingency over 2646 tokens — both 2070, stock-only 81, distilled-only 80, neither 415 → **161 tokens flipped, net −1**. **McNemar exact p = 1.0.**

Distillation genuinely trained: **32 of 146 tensors moved, exactly the `q_proj`/`v_proj` of all 16 layers**, relative Frobenius delta median 5.96e-03. It changed 6.1% of token decisions — **symmetrically**. It moved the draft; it did not move agreement.

So the earlier 0.0 pp was not an artefact of the 360M←135M size inversion. The result is the same at 8B.

## The stronger finding: α was never the lever

Measured with the same kernel: target **39.28 tok/s**, draft **65.90 tok/s** — a 6.5x smaller model is only **1.68x faster**, because at batch 1 both are launch-latency bound. From those measured latencies:

- **break-even acceptance at the shipped default k=5 is 0.832**; measured is 0.813
- a **perfect draft (α = 1.0)** caps at **1.507x** ideal at k=5, and framework overhead costs a further **1.99x** on top
- `k=5` is mis-tuned here: at the measured α the ideal peaks at **k=1–2 (~1.13x)**, not k=5 (0.955x)

No achievable acceptance rate makes this pair pay. That is a stronger statement than "acceptance didn't move", and it means the feature's headline claim needs scoping to pairs where the draft is genuinely cheap relative to the target.

## The pair this issue proposed cannot be run by the shipped feature

Qwen2.5-32B ← 0.5B is refused, because Qwen2.5 large models declare `config.vocab_size` 152064 and small ones 151936. Worse, the two subcommands **disagree**: `distill` refuses up front (correct), while `measure` accepts, does all the work, then dies inside transformers and **discards every completed measurement**. Filed as #344.

Salvaged by calling the same shipped kernels with the error caught: acceptance **0.6857** (MODERATE), plain 25.54 tok/s. The assisted arm is not obtainable.

## A metric bug found along the way — #345

`measure_acceptance` scores the draft against **penalised** target logits, so a draft identical to the target scores **0.8623** on Qwen instead of 1.0. Precision was tested and rejected as the cause. It flips the verdict band on the 32B (0.6857 MODERATE → 0.7072 STRONG), so a CI gate at `--min-acceptance 0.70` would fail that pair for an artefact. Real speculative decoding is unaffected — assisted generation passes the same processors to the draft.

## Not covered
The other scenario this issue mentions: a **genuinely diverged, hard-fine-tuned target**. Arm B is a same-family stock pair, so it answers "was the null result just size?" (no) and not "does distillation help when the target has drifted from the draft?". That remains open and would need its own run.

Record: `benchmarks/gate-h100-validation.md` STEP 26.
